src/components/modebar/manage.js

```diff
@@ -72,8 +72,10 @@ function getButtonGroups(gd) {
     var fullLayout = gd._fullLayout;
     var fullData = gd._fullData;
     var context = gd._context;
-    var buttonsToRemove = context.modeBarButtonsToRemove;
-    var buttonsToAdd = context.modeBarButtonsToAdd;
+    var buttonsToAdd = context.modeBarButtonsToAdd
+        .concat(fullLayout.modebar.add.split('+'));
+    var buttonsToRemove = context.modeBarButtonsToRemove
+        .concat(fullLayout.modebar.remove.split('+'));
```

Comment:
Trying to think if there are any cases we need to explicitly deal with precedence of config over layout. That's the way we decided it should work, right?

If config adds a button that layout has asked to remove, it can add it via the full object, but the string versions aren't defined in the add section below.

And if config removes a button that layout has asked to add, since removal happens first it won't be available yet when it's expected to be removed, so it will appear anyway.

So it seems like these `concat` need to filter out layout items that are in the opposite context entry.

Reply:
Good call. Addressed in https://github.com/plotly/plotly.js/pull/5660/commits/3daa2f3ddb9a4d5cd2a9f6b054f59f8206a929db.